processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,44 +58,107 @@
  * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP or other protocols.
  * It does periodic snap shot of the list of lookup in order to bootstrap nodes after restart.
  */
-
 @ManageLifecycle
 public class LookupReferencesManager
 {
-  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
-  private final ConcurrentMap<String, LookupExtractorFactory> lookupMap = new ConcurrentHashMap<>();
-  // This is a lock against the state of the REFERENCE MANAGER (aka start/stop state), NOT of the lookup itself.
+  private static final EmittingLogger LOG = new EmittingLogger(LookupReferencesManager.class);
+
   private final ReadWriteLock startStopLock = new ReentrantReadWriteLock(true);
-  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @VisibleForTesting
+  volatile boolean started = false;
+
+  @GuardedBy("startStopLock")
+  private final Map<String, LookupExtractorFactoryContainer> lookupMap = new HashMap<>();
 
   private final LookupSnapshotTaker lookupSnapshotTaker;
 
+  @VisibleForTesting
+  final BlockingQueue<Notice> queue = new ArrayBlockingQueue<>(10000);
+
+  private volatile ExecutorService exec;
+
+  //for unit testing only
+  private final boolean testMode;
+
   @Inject
-  public LookupReferencesManager(LookupConfig lookupConfig, final @Json ObjectMapper objectMapper)
+  public LookupReferencesManager(LookupConfig lookupConfig, @Json ObjectMapper objectMapper)
+  {
+    this(lookupConfig, objectMapper, false);
+  }
+
+  @VisibleForTesting
+  LookupReferencesManager(LookupConfig lookupConfig, ObjectMapper objectMapper, boolean testMode)
   {
     if (Strings.isNullOrEmpty(lookupConfig.getSnapshotWorkingDir())) {
       this.lookupSnapshotTaker = null;
     } else {
       this.lookupSnapshotTaker = new LookupSnapshotTaker(objectMapper, lookupConfig.getSnapshotWorkingDir());
     }
+    this.testMode = testMode;
   }
 
   @LifecycleStart
   public void start()
   {
     startStopLock.writeLock().lock();
+
     try {
-      if (!started.getAndSet(true)) {
+      if (!started) {
+        LOG.info("LookupReferencesManager is starting.");
+
         if (lookupSnapshotTaker != null) {
           final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
           for (LookupBean lookupBean : lookupBeanList) {
-            this.put(lookupBean.name, lookupBean.factory);
+            LookupExtractorFactoryContainer container = lookupBean.getContainer();
+
+            if (container.getLookupExtractorFactory().start()) {
+              lookupMap.put(lookupBean.getName(), container);
+            } else {
+              throw new ISE("Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
+            }
           }
         }
-        LOGGER.info("Started lookup factory references manager");
+
+        if (exec == null && !testMode) {
+          exec = Execs.singleThreaded("lookup-reference-manager-%d");
```

Comment:
this *LOOKS* like you're allowing multiple starts and stops of the manager. But there is no epoch tracking, so whenever the handlers are running they can accidentally start processing even though they were submitted from a prior epoch.

I think the safest thing here in general is to have the shutdownNow kick an interrupt to all the waiting read locks, which means making all the read locks in the Notice impls acquire interruptibly. BUT then you have a scenario where you had a thing that looks like it was accepted by the node that was actually NOT accepted by the node because it never started.